### PR TITLE
Augment haskell hangman

### PR DESCRIPTION
**Pull request by SkyBirdSoar**

### CODE REVIEW

**Comment by YorickPeterse** [comment]
Was my Pandoc based Haskell hangman too hard? Other than that, no, I don't mind.

File: bot/src/plugins/hangman.cr [hunk 1]
```diff
@@ -10,22 +10,22 @@ class Hangman
...
+      "gems"           => word_list_for("gem_names"),
+      "german"         => word_list_for("wordlist_de"),
+      "git"            => word_list_for("git"),
+      "haskell"        => word_list_for("haskell_prelude"),
+      "java"           => word_list_for("java_classes"),
+      "nouns"          => word_list_for("wordlist"),
+      "perl"           => word_list_for("perl"),
+      "php"            => word_list_for("php"),
+      "ruby"           => word_list_for("ruby")
     }
```
Comment (jhass) [inline bot/src/plugins/hangman.cr]: No need for the alignment changes here.

File: bot/res/haskell [hunk 2]
```diff
@@ -1,3574 +0,0 @@
...
-Yaml.Value
-Yaml.YamlMark
-YCbCrCoefficients
-YCbCrPositioning
-You
-YResolution
-YYYY
-Z0
-Zone
-ZonedTime
```
Comment (jhass) [inline bot/res/haskell]: So what's wrong with these?

**Comment by SkyBirdSoar** [comment]
Changes:
1. Add haskell_prelude in res/
2. Remove @yorickpeterse's hangman from existence
3. Update hangman.cr

@yorickpeterse's hangman was too difficult and very broad in scope, including things that might not have shipped with haskell.
To support my claim, I shall quote @adaedra:
> 21:47:21 adaedra: How are we even supposed to guess things like A97C

File: bot/res/haskell (hunk 2 above)
Comment (jhass) [inline bot/res/haskell]: > yorickpeterse's hangman was too difficult and very broad in scope, including things that might not have shipped with haskell.
To support my claim, I shall quote adaedra:
> 21:47:21 adaedra: How are we even supposed to guess things like A97C

Discovering new things is pretty much the entire point of the game. Always winning isn't.

Comment (SkyBirdSoar) [inline bot/res/haskell]: Are these really useful to discover? Would you immediately think of using A97C to represent your data? Datatypes are abstractions to represent data and are usually very specific to the situation. Ususally, the constructors are hidden from you. The most you can do is to use methods from that module to get the type and use it in your type signatures. The list does not include those methods, even. Hence, the benefits of using this list for hangman is greatly minimised.

Compare this with my list, the Haskell Prelude, which defines most of the most commonly used methods in Haskell. For example, zip. Zip is a common idiom in Haskell to pair elements from two lists together. Usually, people use list comprehension or map to achieve the result, and those methods are less concise. Hence, by using this list, people will find out about zip, a common idiom, to help them reduce duplication, reinventing the wheel and work, which ultimately benefits them.

This is why this change is better/necessary to remain true to the objectives of the game.

Comment (jhass) [inline bot/res/haskell]: Why not have both?

**Comment by adaedra** [comment]
Please note that what I'm throwing randomly on IRC may not be my definitive, pondered way of thinking.

I actually have no knowledge of Haskell, and, like in other lists, including on ones I introduced myself, I sometimes stop to the random guessing or let other play. I don't know what `A97C` – and actually, don't care about it. I can't say if it would be easy to guess for an intermediate or expert Haskell-person. I don't know if numbers are common in Haskell names. I'm a Haskell John Snow. The sentence you quote comes from my ignorant point of view and may not suit one of someone who cares more or think differently.

Like @jhass said, the game is also discovering, and not always winning. There could be other ways of enhancing the game to help a little on difficult cases or anything else – this is, I think, not the scope of this issue, though. I'm not sure removing too hard guesses is the best solution. I'd say it should be discussed in `#ruby-offtopic` or in a separate issue before blindly removing or maybe have different opinions from common hanged people.

Even if, of course, @jhass is the king here and he will have the final word.

**Comment by SkyBirdSoar** [comment]
Here is my stand. At first, I wanted to preserve @YorickPeterse's list, letting it coexist along with my list. Now, after reading the first `800 lines`, I don't think so.

Firstly, there is little value in the first 800 lines. Let me explain:

First off, it puts me, and maybe others, like @adaedra (he didn't say this!), off. Why is this so? It is insanely difficult and there is no chance of winning, and I see no point of playing it because frankly, I don't know what the constructors are for. `A1`? What's that? There is little chance for me to ever use it at all. Ever.

Why? The reason is because this list is generated for pandoc. The list is really specific to pandoc. How would I remember what `A1` is? `Applicative1`? `Aeson1`? `Amortized1`? We do not have ruby hangman that is specific to a dsl. It is not useful for haskell hangman, generally speaking. Examples: `BeginAfter`, `BlockIndented`, `BlockQuotation`, `ContentRow`, `DestinationURL`, `DokuWiki`. Even worse, they are just types that can not be manipulated, they are only there for type safety (to the best of my knowledge). At most, I suggest renaming it to pandoc.

Numbers arent really common when defining haskell types. `Char8`, `Word8` are just a few I can think of. Hence, this list is a misrepresentation. Well, who would want to learn a language that has so many numbers for you to guess? `A97C` is definitely difficult, if not impossible, to guess. `A1` to `A9` -> Boy is it impossible to find documentation for this, or maybe I didn't try hard enough at google.

It includes the stuff in my list, plus a load of unnecessary things. (Because above).

Okay, so now you know why I wanted to replace it. But why is my list better?

Firstly, Prelude is easy, it was meant to get people into wanting to learn Haskell.

Secondly, it includes the `200 words` that are useful to know (it's like knowing ruby core) and are common abstractions to help you write less code. Compare that with `3500 words` like `B8`? Even if you are not a haskell programmer, the functions defined in Prelude are usually found in most functional languages. You can search for such methods in Ruby as well. Let me provide and example.

```ruby
# Imperative
def transform_array_to_something_else my_array
  temp_array = []
  my_array.each do |item|
    temp_array << do_something item # example, item.to_i
  end
  temp_array
end
```

Upon discovering `map` in Haskell:

```ruby
my_array.map do |item|
  do_something item
end

# or more concisely using the example above,
my_array.map &:to_i
```

Therefore, there are benefits to including functions in Prelude.

To further support my claim, functions such as `foldl` are often not used appropriately (memory leaks), the companion version `foldl'` is often used in its place. A quick google search on the comparisons of `foldl` and `foldl'` will help improve anyone's code. Another example: `sequence`. Sometimes you just want to discard the result. There is `sequence_` for that. No need to write extraneous code.

Furthermore, constructors provided by Haskell is provided. Eg `Just a`. Wow! Something that is very useful for your code! If I do include `lens`, I am sure there would be some very very challenging constructors, I heard it's `s t a b` or something (heh).

*Planned Improvements*
In case those weren't enough for you, I plan to added improvements now.
- `Language Extensions` no one knows them all, it's good to discover them in order to do what you want in the future! I bet you didn't know `NamedFieldPuns` would come in useful when dealing with records!

- `Symbol operators` - although removed for now, I am planning to include them while making them easier as Haskell relies a lot on symbols. Eg. Turning `(*)` to "Multiplies 2 arguments, *". It might seem too easy but there are higher leveled operators in Haskell, eg `<|>`, `<$>`, `<*>`, `$>`, `*>`, `&&&`, `***`. Not all of them are in the Prelude, however, but the list can be expanded to include them.

- *Adding more useful functions and constructors` -
How would i like to do this? I'm going to throw in `Data.List`, `Control.Monad` and others as well to raise the level of difficulty. For example, Data.List includes `intersperse` and `isPrefixOf` which are useful functions i have found during my use of Haskell. `Control.Monad.Error` is also useful for error handling, with constructors like `EitherT`, `ErrorT`, `catch`, `handle` and so on.

That's the end of my argument.

I suggest that we have further discussion upon the original list's use in a separate issue. I did not intend to use this pr to discuss about its usefulness.

I am also going to close this pr and fix the alignments.

I would like to thank you folks for
- `git rebase` - it was good learning experience for me to resolve merge conflicts.
- giving me the opportunity to defend my opinions

And finally, sorry @yorickpeterse (and the pandoc team (please do not tell them about this))

TL;DR
- closing pr
- future
  - new issue on @yorickpeterse
  - new pr
  - new issue on improvements to the list

To make this a even better proposition, I can stay off haskell hangman (we all know how much I like to play hangman) to let others play & potentially maintain haskell hangman (no guarantees!).
